### PR TITLE
Fix (#1046) : Stops fetch request by agency polluted by non-requested agencies stops.

### PR DESCRIPTION
**Pull request by yash170603**
The pr tries to solve #1046 
In multi-agency GTFS feeds, BuildRouteReferences (in internal/restapi/reference_utils.go) calls
  GetRoutesByIDs without filtering by agency_id. This causes incorrect behavior where routes matching the
  ID from any agency in the feed are included in the results, rather than only those belonging to the
  targeted agency.

  Propsed fix:-
  Updated BuildRouteReferences to incorporate agency_id filtering when querying for routes, ensuring that
  route references are correctly scoped to the requested agency.

## Bug Fixes

* Preserved agency information in route references.
* Improved handling of unqualified or malformed route IDs.
* Prevented duplicate route references.
* Ensured route details use the correct agency and handle missing values safely.
* Improved cancellation handling during route lookups.
* Returned consistent results for empty route-reference requests.
* Kept route results correctly scoped across agencies.
* Included all relevant agency references when routes are served by multiple agencies.

### CODE REVIEW

File: internal/restapi/reference_utils_test.go [hunk 1]
```diff
@@ -258,3 +259,116 @@ func TestBuildStopReferencesAndRouteIDsForStops_DeduplicatesStopIDs(t *testing.T
...
+		stops := []models.Stop{
+			{ID: "A1_s1", StaticRouteIDs: []string{"A1_r100", "A1_r200"}},
+			{ID: "A1_s2", StaticRouteIDs: []string{"A1_r100", "A1_r200"}},
+		}
+
+		routes, err := api.BuildRouteReferences(ctx, "A1", stops)
+		require.NoError(t, err)
+		assert.Len(t, routes, 2, "Duplicate route references across multiple stops must be deduplicated")
+	})
+}
```
[the remark this answers is not in the document]
Comment (yash170603) [reply internal/restapi/reference_utils_test.go:400]: Added the requested tests

File: internal/restapi/reference_utils_test.go [hunk 2]
```diff
@@ -258,3 +259,116 @@ func TestBuildStopReferencesAndRouteIDsForStops_DeduplicatesStopIDs(t *testing.T
...
+	// and Agency "A2" has routes r300, r400.
+	multiAgencyFiles := map[string]string{
+		"agency.txt": "agency_id,agency_name,agency_url,agency_timezone\n" +
+			"A1,Agency One,http://agency1.com,America/Los_Angeles\n" +
+			"A2,Agency Two,http://agency2.com,America/Los_Angeles\n",
+		"routes.txt": "route_id,agency_id,route_short_name,route_long_name,route_type\n" +
+			"r100,A1,100-A1,Route 100 For Agency 1,3\n" +
+			"r200,A1,200-A1,Route 200 For Agency 1,3\n" +
+			"r300,A2,300-A2,Route 300 For Agency 2,3\n" +
+			"r400,A2,400-A2,Route 400 For Agency 2,3\n",
```
[the remark this answers is not in the document]
Comment (yash170603) [reply internal/restapi/reference_utils_test.go:308]: Updated the test as requested

**Comment by Ahmedhossamdev** [comment]
Hey @yash170603, Could you please fix the format issue and [sonarqubecloud](https://github.com/apps/sonarqubecloud) issue?
image

**Comment by yash170603** [comment]
Hey @Ahmedhossamdev, thanks for the review! Here is the update on both items:

### Format issue
Fixed — removed the trailing blank line at end of `reference_utils_test.go` and added doc comments to all functions touched by this diff. CodeRabbit should re-scan and pass the 80% docstring threshold on the next run.

### SonarQubeCloud 0.0% Coverage on New Code
This is the standard result for **all** fork PRs — SonarCloud cannot read the `SONAR_TOKEN` secret from forks, so coverage data is never uploaded. You can verify this on every recently merged fork PR:

- PR #1359 (merged) — `0.0% Coverage on New Code`
- PR #1360 (merged) — `0.0% Coverage on New Code`

The CI already generates `clean_profile.cov` correctly (`.github/workflows/go.yml:64-80`), and `sonar-project.properties` points to it. The scan simply never runs for fork PRs because `SONAR_CI_ANALYSIS_ENABLED` is `false` at line 41. Coverage is not blocking the quality gate.

### CodeRabbit: collision test (same route ID across agencies)
CodeRabbit suggested adding a test where agency A2 also has `route_id=r100`. This is **impossible** with the current schema — `routes.id` is `PRIMARY KEY`:

```sql
CREATE TABLE routes (
    id TEXT PRIMARY KEY,        -- unique constraint
    agency_id TEXT NOT NULL,    -- NOT NULL
    FOREIGN KEY (agency_id) REFERENCES agencies(id)  -- FK enforced
) STRICT;
```

I verified this against an in-memory SQLite instance:
- `INSERT INTO routes VALUES('r100','A1',...)` — succeeds
- `INSERT INTO routes VALUES('r100','A2',...)` — **`UNIQUE constraint failed: routes.id`**

GTFS spec also mandates unique `route_id` values per feed. Two agencies cannot share the same `route_id` in a single feed.

### CodeRabbit: empty AgencyID fallback test
Also impossible — the schema enforces both `NOT NULL` and a foreign key:
- `INSERT INTO routes VALUES('r','','...')` — **`FOREIGN KEY constraint failed`** (empty string is not a valid agency)
- `INSERT INTO routes VALUES('r','NONEXISTENT',...)` — **`FOREIGN KEY constraint failed`**

The empty-AgencyID fallback at `reference_utils.go:93-95` is defensive dead code — it can never be triggered through normal operation because the DB rejects it at insert time.

### Summary
- Format fixed
- Docstrings added (83%+ coverage)
- Tests cover all schema-possible edge cases
- SonarQube 0.0% — standard for fork PRs, does not block merge

**Review by burma-shave — CHANGES_REQUESTED** [review]
I think this PR is moving in the right direction, but I’m requesting changes because it only fixes half of the legacy/reference contract.

Looking at legacy Java, route references are hydrated from the routes already exposed by each stop: `BeanFactoryV2.getStop()` adds each route ID to the stop and calls `addToReferences(route)`. `getRoute(route)` then preserves the route’s own agency and also calls `addToReferences(route.getAgency())`.

So I don’t think `BuildRouteReferences` should be a strict “requested agency only” filter. Its contract should be: for every route ID already exposed by the stops, add the matching route reference using that route’s real agency. This PR mostly does that.

The remaining gap is that the callers still usually populate `data.references.agencies` with only the requested/current agency. After this change, `data.references.routes` can include a cross-agency route such as `{ id: "A2_r300", agencyId: "A2" }`, but `data.references.agencies` may still omit `A2`. That leaves the route’s `agencyId` unresolved in the references block, whereas legacy’s `addToReferences(route)` would include the owning agency too.

Could you update the PR so every agency owning a returned route reference is also included in `references.agencies`, and add an endpoint-level test for a cross-agency stop/route case? The helper-level tests are useful, but the response-level invariant we need to protect is:

> every `references.routes[].agencyId` has a matching agency in `references.agencies`.

Once that invariant is covered, this should line up much better with both #1046 and the planned multi-agency support direction.

**Comment by yash170603** [comment]
@burma-shave , sure will pick this up and fix it

**Comment by yash170603** [comment]
Hey @burma-shave,  you're right that this was only half done. Pushed a fix.

Now after stopsForAgencyHandler builds the route references, I loop through them and if a route belongs to a different agency than the one requested, I add that agency into references.agencies too (reused the appendRouteAgencyReference helper that's already used in trip_for_vehicle_handler.go, so it's not a new pattern).

Also added an endpoint-level test like you asked — sets up a stop for A1 that's also served by an A2 route, hits the actual handler, and checks that every route in the response has its agency present in references.agencies. Not just checking the one example, it checks all of them generically.

To be sure this actually does something (didn't want to just add a test that passes no matter what), I reverted the fix and reran it — it failed with basically the exact gap you described, route agency missing from the agencies list. Put the fix back, ran the whole test suite, all green.


One thing — you said "callers" plural, I only fixed stopsForAgencyHandler since that's what the issue is about. stops_for_route_handler.go has the same problem though. Should I fix that here only? or a different fix to avoid polluting the pr.!

Thanks

**Review by burma-shave — CHANGES_REQUESTED** [review]
Thanks for the update — the new `stops-for-agency` endpoint-level test and handler change address the specific gap I called out there.

I agree with your note that `stops_for_route_handler.go` has the same issue. Since this PR changes `BuildRouteReferences` from a caller-agency relabeling helper into a true route-reference hydrator that preserves each route’s owning agency, all direct callers that publish those route refs need to preserve the same response invariant:

> every `references.routes[].agencyId` must have a matching agency in `references.agencies`.

Right now `stops-for-agency` now does that, but `stops-for-route` still builds route refs through `BuildRouteReferences` and only seeds `references.agencies` with the queried route’s agency. A shared stop served by another agency’s route can therefore return a route reference whose `agencyId` is not present in `references.agencies`.

Could you please extend this PR to cover the `stops-for-route` call site too, with a small endpoint-level test for the same cross-agency stop/route response invariant? That keeps the shared helper’s new contract consistent at its other directly affected endpoint.

`trip-details` may also have a similar schedule-reference path

**Comment by yash170603** [comment]
@burma-shave , sure I will pick this up!

**Review by burma-shave — APPROVED** [review]
